lib/server/controllers/file_controller.rb

```diff
@@ -104,4 +107,91 @@ def copy
     )
     success_json(files: [ new_file.to_hash(@request) ])
   end
+
+  def bulk_copy
+    # We want to support bulk copy operations
+    #   from Magma.
+    # First, check that the user has access to all of
+    #   the revision buckets and files, and that
+    #   all revisions are valid.
+    # If all revisions are valid, execute them.
+    require_param(:revisions)
+
+    revisions = @params[:revisions]
+
+    raise Etna::BadRequest, 'At least one revision required' unless revisions.length > 0
+
+    raise Etna::BadRequest, 'All revisions require "source" parameter' unless revisions.all? {|revision| revision.has_key? :source}
+    raise Etna::BadRequest, 'All revisions require "dest" parameter' unless revisions.all? {|revision| revision.has_key? :dest}
+
+    all_source_buckets = revisions.select {|rev| rev[:source]}.
+      map {|rev| extract_bucket_from_path(rev[:source])}.uniq
+    all_dest_buckets = revisions.select {|rev| rev[:dest]}.
+      map {|rev| extract_bucket_from_path(rev[:dest])}.uniq
+
+    user_authorized_bucket_names = Metis::Bucket.where(
+      project_name: @params[:project_name]
+    ).all.select{|b| b.allowed?(@user, @request.env['etna.hmac'])}.
+      map {|bucket| bucket.name}
```

Comment:
It might be easier to break down `Metis::Bucket#allowed?` here, and just extract the `id` (if any) from the valid hmac. Then you could retrieve all buckets `.where(owner: [ 'metis', id ], name: all_source_buckets + all_dest_buckets)`.

Reply:
Would that do the same ROLES check, if I skip `allowed?`? Seems like this would be actually more restrictive, right, only letting users copy to / from buckets where they are owners?

Reply:
You're right, it wouldn't run those ROLES checks, and doing those checks in a db call is hard, so you'd have to use `allowed?` anyway, but adding in the owner and name requirements might let you pare down the number of buckets you are checking.